### DISCUSSION TITLE
Is it possible to download a file without having a torrent or magnet?

### DISCUSSION THREAD

**Discussion by ghost** [arvidn/libtorrent#7677] [General]
Hi,

I'm writing a [tool](https://gitlab.com/NamingThingsIsHard/linux/nix-subsubd) that's supposed to download a file using a content address generated from just a SHA256 hashsum of the file. The original idea was to use IPFS, but it has a [bug](https://github.com/ipfs/kubo/issues/10426) preventing it from doing so.

Here I am now wondering, is there some way to download a file from a swarm just by knowing the SHA256 hashsum? Maybe the DHT or querying some tracker? Or maybe libtorrent already provides a function to do so?

**Comment by master255** [comment]
Unfortunately, both systems won't work for this.
Torrents - has a DHT that contains hash sums of .torrent files.
Ipfs - has a DHT that contains hash sums of chunks of files.

But there is good news. There is Libp2p with which you can create your own DHT in which you will publish hashes of whole files. And you will get your torrents where you can get files by their hashes.
But do you need it? Think about it.

It's simple.

**Comment by arvidn** [comment]
there is no direct facility for this now. However, there's been talk about a convention where clients announce "pretend" single-file v2-torrents, with some canonical piece size, for large files of torrents. that way one could produce the same torrent just based on the file content and discover other torrents containing that file. I can't dig up the thread right now though.

**Reply by ghost** [reply]
Thanks. That would be a great way forward. 

From https://blog.libtorrent.org/2020/09/bittorrent-v2/ does mention that the root  hash of the tree will be (is?) stored in v2 torrents. Would that still allow creating a torrent just from a hashsum of a file? I'm guessing not because the file's hashsum and the root hash _should_ be different  :eyes:

**Reply by arvidn** [reply]
the "root hash" refers to the root of a merkle tree for a specific file in the torrent. Identical files will always produce the same root hash. If you have two v2 torrents that contain the same file, both will have the same root hash.

I don't know what you mean by "hashsum", but if you mean the info-hash, that is something different. And it's less likely to be deterministic since it's the hash of a bencoded structure that (in theory) could contain custom fields.

The v2 specification says that `.torrent` files must contain the "piece layers" as well. i.e. the merkle tree nodes at the piece size of each file. So, without those (or the content of the file so you can compute them) you can't technically create a .torrent file. However, you could produce a magnet link. The magnet link wouldn't work unless there is at least one peer that has the piece layers though.

**Reply by ghost** [reply]
By hashsum I mean the output of SHA256/MD5/blake/etc. The are still many sites that provide SHA256 sums to verify downloads. From your answer, can I correctly infer those (SHA256 sums) will not be enough solely to be able to reconstruct even a magnet link? One would require a mapping of SHA256sum to root hash in order to generate magnet links.

Since I have your attention, one related question: which hash function is used for the hash tree / merkle tree? Unless I missed it, the [spec](https://www.bittorrent.org/beps/bep_0052.html) doesn't seem to indicate that. Or was BLAKE3 just used since it itself uses a hash-tree?
